### PR TITLE
Resource based caching

### PR DESCRIPTION
**Pull request by mpsijm**
Fixing @jasperdenkers' `TODO` comments, while also performing some cleanup tasks :slightly_smiling_face:

### CODE REVIEW

File: org.spoofax.jsglr2/src/main/java/org/spoofax/jsglr2/imploder/incremental/IncrementalTreeImploder.java [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
     private final IIncrementalImplodeInputFactory<ParseNode, Tree, Input> incrementalInputFactory;
     private final TreeImploder<ParseForest, ParseNode, Derivation, Tree, Input> regularImplode;
-    private final Map<String, WeakHashMap<ParseNode, SubTree<Tree>>> cache = new HashMap<>();
+    private final Map<FileObject, WeakHashMap<ParseNode, SubTree<Tree>>> cache = new HashMap<>();
```
Comment (hendrikvanantwerpen) [inline org.spoofax.jsglr2/src/main/java/org/spoofax/jsglr2/imploder/incremental/IncrementalTreeImploder.java]: @mpsijm: @Gohla told me that `FileObject` does not guarantee `equals`/`hashCode` based on the file it represents. He said always to use `FileName` for keys in maps.

Comment (mpsijm) [reply org.spoofax.jsglr2/src/main/java/org/spoofax/jsglr2/imploder/incremental/IncrementalTreeImploder.java]: Thanks for catching this, I did not know that! Will fix :slightly_smiling_face:

Comment (jasperdenkers) [reply org.spoofax.jsglr2/src/main/java/org/spoofax/jsglr2/imploder/incremental/IncrementalTreeImploder.java]: Thanks @hendrikvanantwerpen.

Reference that confirms @mpsijm's [fix](https://github.com/metaborg/jsglr/pull/64/commits/501f87cbc93bc5731516cc4e91e8ea69e1e696bf) is safe: ["File names are immutable, and work correctly as keys in hash tables."](https://commons.apache.org/proper/commons-vfs/commons-vfs2/apidocs/org/apache/commons/vfs2/FileName.html).

**Comment by jasperdenkers** [comment]
The reason to start using `FileObject` was for generating [`IMessages`](https://github.com/metaborg/spoofax/blob/master/org.metaborg.core/src/main/java/org/metaborg/core/messages/IMessage.java).

@gohla Does your suggestion imply that the resource field must/could be removed from `IMessage`? An alternative approach is that JSGLR gets its own representation of messages which can be mapped to `IMessage`.

@mpsijm Could you think of how the interface should look like for such an external cache mechanism?

**Comment by mpsijm** [comment]
Reply to @jasperdenkers:

> Does your suggestion imply that the resource field must/could be removed from `IMessage`? An alternative approach is that JSGLR gets its own representation of messages which can be mapped to `IMessage`.

I think that your alternative approach would be best, because that would make JSGLR2 independent of which Spoofax version is used.

> Could you think of how the interface should look like for such an external cache mechanism?

To allow external caching, we should also expose the `IParser`, `IImploder`, and `ITokenizer` components because the intermediate results need to be cached as well. Each of these interfaces will need to have two "main" methods: one for batch parsing (without caching) and one for incremental parsing (where the previous result will need to be manually passed along).
In this system, having a built-in (but replaceable) cache implementation, as @Gohla suggested, seems useful to me.

**Comment by mpsijm** [comment]
@jasperdenkers summary from an offline conversation with @Gohla (thanks btw!):

- To remove the caching responsibility from the parser, imploder, and tokenizer, you'd need to pass the cached values to these components from the outside. In the façade class `JSGLR2`, we can implement these caches with some `HashMap`s, to simplify the interface for external users (and not to have to change anything in Spoofax Core).
- In order to pass cached previous versions to the parser, imploder and tokenizer, they need an extended interface that allows passing these parameters. Here we have two options: (using just the parser as example, but the same applies to the imploder and tokenizer)
  - Have two interfaces: `IParser` and `IIncrementalParser extends IParser`. This enforces at compile-time that you cannot pass caches to non-incremental parsers, at the cost of severely complicating the `ParserVariant` switch that builds a parser based on options. We'd probably also need an `IncrementalParserVariant` class to avoid having to down-cast when a user builds an incremental parser.
  - Have one interface: `IParser`, where the `parse` method has two extra parameters to pass the previous `String` and `ParseForest`. The non-incremental implementations can just ignore these extra parameters (maybe throw a warning somewhere), which is slightly uglier but at least does not require changing `ParserVariant`. Of course, this single interface can have a `default` method that does not have parameters for the cached previous versions and automatically passes `null` for these.

  My preference leans towards the second option, also taking into account @Gohla's remark that in the ideal future, we'll only want to use the incremental version anyway. For batch parsing (e.g. the CLI), users can always leave out the cache parameters.

- After fixing all this, the parser and imploder don't need to know anything about filenames anymore, so we can remove the filename parameter from these two interfaces. At the moment, the tokenizer still needs the filename to attach to tokens because of JSGLR1 legacy and a connection with some Stratego primitives. We can probably rewrite these primitives to get the filename in a different way once JSGLR1 and NaBL1 are completely phased out. Of course, the `JSGLR2` façade will keep the filename parameter to facilitate caching (and, currently, to pass it to the tokenizer).

**Comment by Gohla** [comment]
Adding a little bit to that: JSGLR should not depend on `org.metaborg.core`, etc, as that would make any user of the parser depend on MetaBorg/Spoofax Core, which we want to avoid. This means you do have to duplicate a little bit of code, as you need to create your own `Message` class.

**Comment by jasperdenkers** [comment]
Sounds like a plan 👍 Let's do the proposed changes separately.

**Comment by jasperdenkers** [comment]
Removed [the dependency on `org.metaborg.core`](https://github.com/metaborg/jsglr/commit/a2e2a193f4076a86e42f90b86a25a93c345382f6) and [usage of `FileObjects`](https://github.com/metaborg/jsglr/commit/a97b0445ade03e375f19d57751a1b12447a7e099).

**Comment by jasperdenkers** [comment]
See https://github.com/metaborg/jsglr/pull/66 for @mpsijm's work on making caches external.

**Pull request closed without merging**